pennylane/plugins/expt_tensornet_tf.py

```diff
@@ -0,0 +1,342 @@
+
+    def reset(self):
+        self.res = None
+        self.variables = []
+        super().reset()
+
+    def execution_context(self):
+        self.tape = tf.GradientTape(persistent=True)
+        return self.tape
+
+    def pre_apply(self):
+        super().pre_apply()
+
+        self.op_params = {}
+
+        for operation in self.op_queue:
+            # Copy the operation parameters to the op_params dictionary.
+            # Note that these are the unwrapped parameters, so PennyLane
+            # free parameters will be represented as Variable instances.
+            self.op_params[operation] = operation.params[:]
+
+        # Loop through the free parameter reference dictionary
+        for _, par_dep_list in self.parameters.items():
+            if not par_dep_list:
+                # parameter is not used within circuit
+                v = tf.Variable(0, dtype=self.R_DTYPE)
+                self.variables.append(v)
+                continue
+
+            # get the first parameter dependency for each free parameter
+            first = par_dep_list[0]
+
+            # For the above parameter dependency, get the corresponding
+            # operation parameter variable, and get the numeric value.
+            # Convert the resulting value to a TensorFlow tensor.
+            val = first.op.params[first.par_idx].val
+            mult = first.op.params[first.par_idx].mult
+            v = tf.Variable(val / mult, dtype=self.R_DTYPE)
+
+            # Mark the variable to be watched by the gradient tape,
+            # and append it to the variable list.
+            self.variables.append(v)
+
+            for p in par_dep_list:
+                # Replace the existing Variable free parameter in the op_params dictionary
+                # with the corresponding tf.Variable parameter.
+                # Note that the free parameter might be scaled by the
+                # variable.mult scaling factor.
+                mult = p.op.params[p.par_idx].mult
+                self.op_params[p.op][p.par_idx] = v * mult
+
+        # check that no Variables remain in the op_params dictionary
+        values = [item for sublist in self.op_params.values() for item in sublist]
+        assert not any(
+            isinstance(v, Variable) for v in values
+        ), "A pennylane.Variable instance was not correctly converted to a tf.Variable"
+
+        # flatten the variables list in case of nesting
+        self.variables = tf.nest.flatten(self.variables)
+        self.tape.watch(self.variables)
+
+        for operation in self.op_queue:
+            # Apply each operation, but instead of passing operation.parameters
+            # (which contains the evaluated numeric parameter values),
+            # pass op_params[operation], which contains numeric values
+            # for fixed parameters, and tf.Variable objects for free parameters.
+            super().apply(operation.name, operation.wires, self.op_params[operation])
```

Comment:
What would be the reason behind having the calls to ``apply`` on the ``operations`` here? Is it related to the ``tape`` used in tf?

Reply:
We need to call `apply()` with the new parameter dictionary that uses `tf.Variable`s (`self.op_params`). However, if we allowed the default `Device.apply()` method, it will pass `operation.parameters`, which stores the numeric values.

This is just a way of 'bypassing' the default `apply` method, while also trying to minimize code duplication.